pandas/core/indexes/numeric.py

```diff
@@ -224,6 +225,15 @@ def _assert_safe_casting(cls, data, subarr):
                 raise TypeError('Unsafe NumPy casting, you must '
                                 'explicitly cast')
 
+    def _is_compatible_with_other(self, other):
+        return (
+            super(Int64Index, self)._is_compatible_with_other(other)
+            or (
```

Comment:
this is ofttly complicated, shouldn't this work
```
isinstance(type(self), (Int64Index, ABCRangeIndex)) and isinstance(type(other), (Int64Index, ABCRangeIndex))
```?

Reply:
I don't believe this would work.
Ex.

    >>> self = pd.Int64Index([])
    >>> isinstance(type(self), (Int64Index, ABCRangeIndex))  # since Int64Index is not an instance of Int64Index
    False

The following don't work either:

    >>> isinstance(self, (Int64Index, ABCRangeIndex))  # since this would be True for all instances of subclasses of Int64Index
    >>> type(self) in (Int64Index, ABCRangeIndex)  # since the RangeIndex cases won't match
    >>> type(self) in (Int64Index, RangeIndex)  # This would theoretically work, but I believe might cause a circular ref

Reply:
``any([isinstance(type(obj), (Int64Index, ABCRangeIndex)) for obj in [self, other]])``

Reply:
Just replaced `Int64Index` with `ABCInt64Index` and `any` with `all` (which was essentially your first suggestion) and this worked perfectly.